smartmeter_datacollector/smartmeter/cosem.py

```diff
@@ -19,19 +37,59 @@ class RegisterCosem:
 
 
 class CosemConfig:
-    def __init__(self, id_obis: str, clock_obis: str, register_obis: List[RegisterCosem]) -> None:
-
-        self._id_obis = id_obis
-        self._clock_obis = clock_obis
+    def __init__(self, fallback_id: str, register_obis: List[RegisterCosem]) -> None:
+        self._id: Optional[str] = None
+        self._fallback_id = fallback_id
         self._register_obis = {r.obis: r for r in register_obis}
 
-    @property
-    def id_obis(self) -> str:
-        return self._id_obis
+    def retrieve_id(self, dlms_objects: Dict[str, Any]) -> str:
+        if self._id:
+            return self._id
+
+        id_obis = self._find_id_obis(dlms_objects)
+        if not id_obis:
+            LOGGER.debug("Unable to find ID object. Using fallback ID %s.", self._fallback_id)
+            return self._fallback_id
+
+        id_obj = dlms_objects[id_obis]
+        if not isinstance(id_obj, GXDLMSData):
+            LOGGER.debug("Invalid ID object. Using fallback ID %s.", self._fallback_id)
+            return self._fallback_id
 
-    @property
-    def clock_obis(self) -> str:
-        return self._clock_obis
+        meter_id = id_obj.getValues()[1]
+        if not isinstance(meter_id, str) or len(meter_id) == 0:
+            LOGGER.debug("Invalid ID. Using fallback ID %s.", self._fallback_id)
+            return self._fallback_id
+
+        self._id = meter_id
+        return meter_id
+
+    def retrieve_timestamp(self, dlms_objects: Dict[str, Any]) -> datetime:
+        clock_obj = dlms_objects.get(DEFAULT_CLOCK_OBIS, None)
+        if not clock_obj or not isinstance(clock_obj, GXDLMSClock):
```

Comment:
I would rewrite the logic so that you only have to call `datetime.utcnow()` to avoid code duplication and the risk of someone changing one occurance and forget about the other in the future. And maybe use a different log level than DEBUG...

```python
if clock_obj and isinstance(clock_obj, GXDLMSClock):
    timestamp = self._extract_datetime(clock_obj)
    if timestamp:
        return timestamp

    LOGGER.warn(f"Unable to parse timestamp from {DEFAULT_CLOCK_OBIS}. Using system time.")
else:
    LOGGER.warn(f"No clock object found at {DEFAULT_CLOCK_OBIS}. Using system time.")

return datetime.utcnow
```

Reply:
This was just off the top of my head and more like pseudo code ;-)